cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -90,6 +95,7 @@
     this.cConf = cConf;
     this.datasetFramework = datasetFramework;
     this.upgradeComplete = new AtomicBoolean(false);
+    this.messagePublisher = getMessagingContext().getMessagePublisher();
```

Comment:
Shouldn't cache the message publisher, but instead call the `getMessagePublisher()` method in the same thread that the transaction is executing in.